File: rust/crates/corlinman-channels/src/service.rs
//! `run_qq_channel` — the long-running task that bridges a live OneBot WS
//! session to the gateway's internal chat pipeline.
//!
//! Flow per inbound event:
//!
//! 1. [`OneBotClient`] delivers a decoded [`Event`].
//! 2. Only `Event::Message` survives the filter.
//! 3. [`ChannelRouter::dispatch`] applies keyword / @mention gating and
//!    produces a [`ChatRequest`].
//! 4. A new task is spawned per accepted message so a slow reasoning loop
//!    doesn't block the next inbound event.
//! 5. That task calls [`ChatService::run`], collects every
//!    `TokenDelta`, and on `Done` posts a `send_group_msg` / `send_private_msg`
//!    action back to the OneBot client via the shared action channel.
//!
//! Error handling is deliberately tolerant — this runs for the life of the
//! process and any transient error in one message must not kill the task for
//! every other group.

use std::collections::HashMap;
use std::sync::Arc;

use corlinman_core::config::QqChannelConfig;
use corlinman_core::types::ChatRequest;
use corlinman_gateway_api::{
    ChatService, InternalChatEvent, InternalChatRequest, Message as ApiMessage, Role as ApiRole,
};
use futures::StreamExt;
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;

use crate::qq::message::{
    segments_to_attachments, Action, Event, MessageEvent, MessageSegment, MessageType,
};
use crate::qq::onebot::{OneBotClient, OneBotConfig};
use crate::rate_limit::TokenBucket;
use crate::router::{ChannelRouter, GroupKeywords, RateLimitHook};

/// Parameters the caller (gateway main) passes in. A simple struct so future
/// additions (model overrides per channel, rate limits) don't churn the
/// signature.
pub struct QqChannelParams {
    pub config: QqChannelConfig,
    /// Default chat model; the inbound message carries no model hint today.
    pub model: String,
    /// Shared chat pipeline.
    pub chat_service: Arc<dyn ChatService>,
    /// Observation hook fired each time a message is dropped by a rate-limit
    /// check. Wired to the gateway's
    /// `corlinman_channels_rate_limited_total{channel, reason}` counter in
    /// production; tests leave it `None`.
    pub rate_limit_hook: Option<RateLimitHook>,
}

/// Spawn the QQ channel loop and run until `cancel` fires. Returns `Ok(())`
/// on a clean shutdown; surfaces a `config` error when the WS URL or self_ids
/// are empty (validated caller-side but defended here too).
pub async fn run_qq_channel(
    params: QqChannelParams,
    cancel: CancellationToken,
) -> anyhow::Result<()> {
    let QqChannelParams {
        config,
        model,
        chat_service,
        rate_limit_hook,
    } = params;

    if config.ws_url.is_empty() {
        anyhow::bail!("channels.qq.ws_url is empty");
    }
    if config.self_ids.is_empty() {
        anyhow::bail!("channels.qq.self_ids is empty");
    }

    let access_token = match config.access_token.as_ref() {
        Some(s) => Some(s.resolve()?),
        None => None,
    };

    // Build token buckets up front; `None` on a field = dimension disabled.
    let group_limiter = config
        .rate_limit
        .group_per_min
        .map(|n| Arc::new(TokenBucket::per_minute(n)));
    let sender_limiter = config
        .rate_limit
        .sender_per_min
        .map(|n| Arc::new(TokenBucket::per_minute(n)));

    // GC the live buckets on a child cancel so we stop sweeping at shutdown.
    let gc_cancel = cancel.child_token();
    let _gc_group = group_limiter
        .as_ref()
        .map(|b| b.clone().start_gc(gc_cancel.clone()));
    let _gc_sender = sender_limiter
        .as_ref()
        .map(|b| b.clone().start_gc(gc_cancel.clone()));

    let mut router = ChannelRouter::new(
        keywords_to_router(&config.group_keywords),
        config.self_ids.clone(),
    )
    .with_rate_limits(group_limiter, sender_limiter);
    if let Some(hook) = rate_limit_hook {
        router = router.with_rate_limit_hook(hook);
    }
    let router = Arc::new(router);

    // Channels the OneBotClient and the dispatch loop share.
    let (event_tx, mut event_rx) = mpsc::channel::<Event>(64);
    let (action_tx, action_rx) = mpsc::channel::<Action>(64);

    let client = OneBotClient::new(
        OneBotConfig {
            url: config.ws_url.clone(),
            access_token,
        },
        event_tx,
        action_rx,
    );

    // Hand the OneBot client its own cancel child so both tasks stop together.
    let client_cancel = cancel.child_token();
    let client_handle = tokio::spawn(async move { client.run(client_cancel).await });

    let dispatch_cancel = cancel.child_token();
    let dispatch_handle = {
        let chat_service = chat_service.clone();
        let router = router.clone();
        let action_tx = action_tx.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    biased;
                    _ = dispatch_cancel.cancelled() => break,
                    maybe_ev = event_rx.recv() => {
                        let Some(ev) = maybe_ev else { break; };
                        let Event::Message(msg_ev) = ev else { continue; };
                        let Some(req) = router.dispatch(&msg_ev) else { continue; };

                        // Spawn the chat task so the next inbound event isn't
                        // blocked by a slow reasoning loop.
                        let chat_service = chat_service.clone();
                        let action_tx = action_tx.clone();
                        let model = model.clone();
                        let cancel = dispatch_cancel.child_token();
                        tokio::spawn(async move {
                            if let Err(err) =
                                handle_one(chat_service, req, msg_ev, model, action_tx, cancel)
                                    .await
                            {
                                tracing::warn!(
                                    target: "corlinman.channels.qq",
                                    error = %err,
                                    "qq message dispatch failed"
                                );
                            }
                        });
                    }
                }
            }
        })
    };

    // Wait for shutdown (cancel) or either task finishing unexpectedly.
    tokio::select! {
        _ = cancel.cancelled() => {}
        res = client_handle => {
            if let Err(join_err) = res {
                tracing::warn!(
                    target: "corlinman.channels.qq",
                    error = %join_err,
                    "onebot client task panicked"
                );
            }
        }
    }

    dispatch_handle.abort();
    Ok(())
}

/// Convert the TOML-loaded `HashMap<String, Vec<String>>` to the router's
/// type alias. They're structurally identical — this just names the intent.
fn keywords_to_router(m: &HashMap<String, Vec<String>>) -> GroupKeywords {
    m.clone()
}

async fn handle_one(
    chat_service: Arc<dyn ChatService>,
    req: ChatRequest,
    event: MessageEvent,
    model: String,
    action_tx: mpsc::Sender<Action>,
    cancel: CancellationToken,
) -> anyhow::Result<()> {
    let internal = build_internal_request(&req, &event, model);

    let mut stream = chat_service.run(internal, cancel.clone()).await;
    let mut text = String::new();
    let mut had_error: Option<String> = None;
    while let Some(ev) = stream.next().await {
        match ev {
            InternalChatEvent::TokenDelta(t) => text.push_str(&t),
            InternalChatEvent::ToolCall { .. } => {
                // Informational — the gateway handles execution.
            }
            InternalChatEvent::Done { .. } => break,
            InternalChatEvent::Error(e) => {
                had_error = Some(e.message);
                break;
            }
        }
    }

    let body = if let Some(err) = had_error {
        // Surface errors as a short system reply so the user knows something
        // failed; M5 scope doesn't define a prettier UX.
        format!("[corlinman error] {err}")
    } else if text.trim().is_empty() {
        // Nothing to send — silently drop (matches qqBot.js behaviour on empty
        // assistant responses).
        return Ok(());
    } else {
        text
    };

    let action = build_reply_action(&event, &body);
    action_tx
        .send(action)
        .await
        .map_err(|e| anyhow::anyhow!("action channel closed: {e}"))?;
    Ok(())
}

/// Convert a routed [`ChatRequest`] + the original [`MessageEvent`] into the
/// `InternalChatRequest` handed to the chat service.
///
/// Attachments are derived from the raw QQ segments here rather than on the
/// router to keep `ChatRequest` purely text; that way schedulers and other
/// non-multimodal callers don't need to thread empty vectors everywhere.
fn build_internal_request(
    req: &ChatRequest,
    event: &MessageEvent,
    model: String,
) -> InternalChatRequest {
    let attachments = segments_to_attachments(&event.message);
    InternalChatRequest {
        model,
        messages: vec![ApiMessage {
            role: ApiRole::User,
            content: req.content.clone(),
        }],
        session_key: req.session_key.clone(),
        stream: true,
        max_tokens: None,
        temperature: None,
        attachments,
        // Backfill the transport binding so gateway-side consumers
        // (context_assembler, approval scoping, daily-note tagging) can reason
        // about provenance without re-deriving it from the session_key.
        binding: Some(req.binding.clone()),
    }
}

/// Build a `send_group_msg` / `send_private_msg` action carrying a single
/// text segment. Group messages prepend an `@sender` so the reply is clearly
/// addressed (matches qqBot.js behaviour for keyword-triggered responses).
fn build_reply_action(event: &MessageEvent, body: &str) -> Action {
    match event.message_type {
        MessageType::Group => {
            let gid = event.group_id.unwrap_or_default();
            let segments = vec![
                MessageSegment::at(event.user_id.to_string()),
                MessageSegment::text(format!(" {body}")),
            ];
            Action::SendGroupMsg {
                group_id: gid,
                message: segments,
            }
        }
        MessageType::Private => Action::SendPrivateMsg {
            user_id: event.user_id,
            message: vec![MessageSegment::text(body.to_string())],
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::qq::message::{KnownSegment, MessageSegment};
    use corlinman_core::channel_binding::ChannelBinding;
    use corlinman_gateway_api::AttachmentKind;

    fn sample_group_event() -> MessageEvent {
        MessageEvent {
            self_id: 100,
            message_type: MessageType::Group,
            sub_type: Some("normal".into()),
            group_id: Some(12345),
            user_id: 555,
            message_id: 42,
            message: vec![MessageSegment::text("格兰早")],
            raw_message: "格兰早".into(),
            time: 1700000000,
            sender: None,
        }
    }

    #[test]
    fn group_reply_addresses_sender() {
        let ev = sample_group_event();
        let a = build_reply_action(&ev, "hello");
        match a {
            Action::SendGroupMsg { group_id, message } => {
                assert_eq!(group_id, 12345);
                assert_eq!(message.len(), 2);
                // first segment is an At targeting user 555
                match message[0].known() {
                    Some(KnownSegment::At { qq }) => assert_eq!(qq, "555"),
                    other => panic!("expected At, got {other:?}"),
                }
                match message[1].known() {
                    Some(KnownSegment::Text { text }) => assert!(text.contains("hello")),
                    other => panic!("expected Text, got {other:?}"),
                }
            }
            other => panic!("expected SendGroupMsg, got {other:?}"),
        }
    }

    #[test]
    fn dispatch_propagates_attachments() {
        let mut ev = sample_group_event();
        ev.message = vec![
            MessageSegment::text("look at this "),
            MessageSegment::Known(KnownSegment::Image {
                url: "https://cdn/pic.png".into(),
                file: Some("pic.png".into()),
            }),
        ];
        ev.raw_message = "look at this [CQ:image,url=https://cdn/pic.png]".into();

        let binding = ChannelBinding::qq_group(ev.self_id, ev.group_id.unwrap(), ev.user_id);
        let mut req = ChatRequest::new(binding, "look at this".into());
        req.message_id = Some(ev.message_id.to_string());

        let internal = build_internal_request(&req, &ev, "claude-sonnet-4-5".into());

        assert_eq!(internal.attachments.len(), 1);
        assert_eq!(internal.attachments[0].kind, AttachmentKind::Image);
        assert_eq!(
            internal.attachments[0].url.as_deref(),
            Some("https://cdn/pic.png")
        );
        assert_eq!(internal.messages.len(), 1);
        assert_eq!(internal.messages[0].content, "look at this");
    }

    #[test]
    fn dispatch_empty_attachments_when_text_only() {
        let ev = sample_group_event();
        let binding = ChannelBinding::qq_group(ev.self_id, ev.group_id.unwrap(), ev.user_id);
        let req = ChatRequest::new(binding, ev.raw_message.clone());
        let internal = build_internal_request(&req, &ev, "claude-sonnet-4-5".into());
        assert!(internal.attachments.is_empty());
    }

    #[test]
    fn private_reply_omits_at() {
        let mut ev = sample_group_event();
        ev.message_type = MessageType::Private;
        ev.group_id = None;
        let a = build_reply_action(&ev, "hi");
        match a {
            Action::SendPrivateMsg { user_id, message } => {
                assert_eq!(user_id, 555);
                assert_eq!(message.len(), 1);
                match message[0].known() {
                    Some(KnownSegment::Text { text }) => assert_eq!(text, "hi"),
                    other => panic!("expected Text, got {other:?}"),
                }
            }
            other => panic!("expected SendPrivateMsg, got {other:?}"),
        }
    }
}

File: python/packages/corlinman-channels/src/corlinman_channels/service.py
"""Channel orchestration helpers — wire an adapter to a chat backend.

Python port of ``rust/.../service.rs`` + the orchestration in
``rust/.../telegram/service.rs``. Provides two ``run_*_channel``
helpers and a ``ChatServiceLike`` Protocol so the per-channel reply
loops stay structurally symmetric with the Rust crate.

## Flow per inbound event

1. The adapter (``OneBotAdapter`` / ``TelegramAdapter``) delivers a
   normalized :class:`InboundEvent`.
2. The router applies keyword / @mention gating and produces a
   :class:`RoutedRequest` (only OneBot today; the Telegram adapter
   already does its own gating in ``inbound()``).
3. A reply coroutine is spawned per accepted message so a slow
   reasoning loop doesn't block the next inbound event.
4. The coroutine calls ``chat_service.run(...)``, collects every
   ``TokenDelta``, and on ``Done`` posts an outbound action / reply.

## Deliberate deviations

- Rust spawns ``tokio::task`` per accepted message; we use
  ``asyncio.create_task``. Behaviour is equivalent on a single-threaded
  asyncio runtime.
- Rust uses ``mpsc`` reply channels typed to the OneBot ``Action``;
  Python keeps them as ``asyncio.Queue`` with the same wire types
  (``Action`` from :mod:`corlinman_channels.onebot`).
- The Telegram outbound path goes through :class:`TelegramSender`
  rather than a reply channel — the Rust crate does the same as of
  the webhook split, so this is parity, not deviation.
- ``ChatService`` is structural (Protocol) so we can decouple from
  ``corlinman-server`` at module load. Pass any object whose ``run``
  yields ``(role, text)``-shaped events.
"""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from dataclasses import dataclass, field
from typing import Any, Protocol

import httpx

from corlinman_channels.common import InboundEvent
from corlinman_channels.onebot import (
    Action,
    MessageEvent,
    MessageType,
    OneBotAdapter,
    OneBotConfig,
    SendGroupMsg,
    SendPrivateMsg,
    TextSegment,
)
from corlinman_channels.rate_limit import TokenBucket
from corlinman_channels.router import ChannelRouter, GroupKeywords, RoutedRequest
from corlinman_channels.telegram import TelegramAdapter, TelegramConfig
from corlinman_channels.telegram_send import TelegramSender

__all__ = [
    "ChatEventLike",
    "ChatServiceLike",
    "QqChannelParams",
    "TelegramChannelParams",
    "handle_one_qq",
    "handle_one_telegram",
    "run_qq_channel",
    "run_telegram_channel",
]


# ---------------------------------------------------------------------------
# Chat-service protocol — structural, decouples this package from the
# concrete corlinman-server types.
# ---------------------------------------------------------------------------


class ChatEventLike(Protocol):
    """One streamed event from the chat backend. The Rust crate has
    a closed enum (``TokenDelta`` / ``ToolCall`` / ``Done`` /
    ``Error``); we accept any object with a ``kind`` discriminator
    string and optional ``text`` / ``error`` attributes."""

    kind: str
    """``"token_delta"`` | ``"tool_call"`` | ``"done"`` | ``"error"``."""

    text: str
    """For ``token_delta``: the delta string."""

    error: str
    """For ``error``: the error message."""


class ChatServiceLike(Protocol):
    """Minimal chat-service surface the orchestration helpers consume.

    Mirrors ``ChatService::run`` in the Rust gateway-api crate; the
    Python ``ChatService`` Protocol in ``corlinman-server`` happens
    to satisfy this shape (modulo the event field names — see
    :func:`_event_kind`)."""

    async def run(
        self,
        request: Any,
        cancel: asyncio.Event,
    ) -> AsyncIterator[Any]:
        """Run one chat turn. Yields events until done. ``cancel.set()``
        should cause the iterator to terminate ASAP."""
        ...


# ---------------------------------------------------------------------------
# QQ channel
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class QqChannelParams:
    """Parameters for :func:`run_qq_channel`. Mirrors Rust
    ``QqChannelParams`` field-for-field, plus a structural ``config``
    so callers don't need a corlinman-core Python port to construct
    one."""

    config: Any
    """``cfg.channels.qq`` — must expose ``ws_url``, ``self_ids``,
    optional ``access_token``, optional ``group_keywords``, and an
    optional ``rate_limit`` sub-struct with ``group_per_min`` /
    ``sender_per_min``."""

    model: str = ""
    chat_service: ChatServiceLike | None = None
    rate_limit_hook: Any = None
    hook_bus: Any = None


async def run_qq_channel(
    params: QqChannelParams,
    cancel: asyncio.Event,
) -> None:
    """Spawn the QQ channel loop and run until ``cancel`` is set.

    Mirrors the Rust ``run_qq_channel`` function. Raises ``ValueError``
    on missing required config (matches Rust ``anyhow::bail!`` shape).
    """
    cfg = params.config
    ws_url = _attr(cfg, "ws_url", "")
    if not ws_url:
        raise ValueError("channels.qq.ws_url is empty")
    self_ids = list(_attr(cfg, "self_ids", []) or [])
    if not self_ids:
        raise ValueError("channels.qq.self_ids is empty")

    # Token buckets — None on either dimension disables it.
    rate_cfg = _attr(cfg, "rate_limit", None)
    group_limiter: TokenBucket | None = None
    sender_limiter: TokenBucket | None = None
    if rate_cfg is not None:
        gpm = _attr(rate_cfg, "group_per_min", None)
        spm = _attr(rate_cfg, "sender_per_min", None)
        if gpm:
            group_limiter = TokenBucket.per_minute(int(gpm))
        if spm:
            sender_limiter = TokenBucket.per_minute(int(spm))

    # GC sweepers tied to cancel — they exit when the event fires.
    gc_tasks: list[asyncio.Task[None]] = []
    if group_limiter is not None:
        gc_tasks.append(group_limiter.start_gc(cancel))
    if sender_limiter is not None:
        gc_tasks.append(sender_limiter.start_gc(cancel))

    router = ChannelRouter(
        group_keywords=_coerce_keywords(_attr(cfg, "group_keywords", {})),
        self_ids=self_ids,
    ).with_rate_limits(group_limiter, sender_limiter)
    if params.rate_limit_hook is not None:
        router = router.with_rate_limit_hook(params.rate_limit_hook)
    if params.hook_bus is not None:
        router = router.with_hook_bus(params.hook_bus)

    adapter = OneBotAdapter(
        OneBotConfig(
            url=ws_url,
            access_token=_attr(cfg, "access_token", None),
            self_ids=self_ids,
        )
    )

    try:
        async with adapter:
            await _qq_dispatch_loop(adapter, router, params, cancel)
    finally:
        for t in gc_tasks:
            t.cancel()


async def _qq_dispatch_loop(
    adapter: OneBotAdapter,
    router: ChannelRouter,
    params: QqChannelParams,
    cancel: asyncio.Event,
) -> None:
    """Inner loop — reads inbound events and spawns per-message reply
    tasks. Equivalent of the Rust ``tokio::select! { cancelled() / recv() }``
    in ``run_qq_channel``."""
    inbound_iter = adapter.inbound()
    pending: set[asyncio.Task[None]] = set()
    try:
        while not cancel.is_set():
            # Get the next inbound event with a cancel-aware wait.
            ev = await _race_iter_or_cancel(inbound_iter, cancel)
            if ev is None:
                break
            payload = ev.payload
            if not isinstance(payload, MessageEvent):
                continue
            req = router.dispatch(payload)
            if req is None:
                continue
            if params.chat_service is None:
                # No backend wired — drop silently (matches Rust when
                # the gateway opts not to provide one).
                continue
            t = asyncio.create_task(
                handle_one_qq(
                    params.chat_service,
                    req,
                    payload,
                    params.model,
                    adapter,
                    cancel,
                )
            )
            pending.add(t)
            t.add_done_callback(pending.discard)
    finally:
        # Best-effort: cancel any in-flight handlers on shutdown.
        for t in pending:
            t.cancel()


async def handle_one_qq(
    chat_service: ChatServiceLike,
    req: RoutedRequest,
    event: MessageEvent,
    model: str,
    adapter: OneBotAdapter,
    cancel: asyncio.Event,
) -> None:
    """Run one chat turn and post the reply back through the adapter.

    Mirrors Rust ``handle_one`` in ``service.rs``. On error, sends a
    short ``[corlinman error] <msg>`` reply so the user knows
    something failed (matches Rust ``M5`` UX).
    """
    request = _build_internal_request(req, event, model)
    stream = await chat_service.run(request, cancel)
    text_parts: list[str] = []
    error_message: str | None = None
    async for chat_ev in stream:
        kind = _event_kind(chat_ev)
        if kind == "token_delta":
            text_parts.append(getattr(chat_ev, "text", "") or "")
        elif kind == "done":
            break
        elif kind == "error":
            error_message = getattr(chat_ev, "error", "") or getattr(
                chat_ev, "message", ""
            )
            break
        # tool_call → informational; gateway handles execution.

    if error_message is not None:
        body = f"[corlinman error] {error_message}"
    else:
        body = "".join(text_parts)
        if not body.strip():
            return  # Empty assistant reply → silent drop.

    action = _build_reply_action(event, body)
    await adapter.send_action(action)


def _build_internal_request(
    req: RoutedRequest,
    event: MessageEvent,
    model: str,
) -> dict[str, Any]:
    """Build the dict payload handed to ``chat_service.run``. We keep
    it dict-shaped to avoid a hard dep on a Python ``InternalChatRequest``
    type; the corlinman-server tests use a TypedDict-friendly shape so
    this round-trips cleanly."""
    from corlinman_channels.onebot import segments_to_attachments

    attachments = segments_to_attachments(event.message)
    return {
        "model": model,
        "messages": [{"role": "user", "content": req.content}],
        "session_key": req.session_key,
        "stream": True,
        "max_tokens": None,
        "temperature": None,
        "attachments": attachments,
        "binding": req.binding,
    }


def _build_reply_action(event: MessageEvent, body: str) -> Action:
    """Build a ``SendGroupMsg`` / ``SendPrivateMsg`` action with a
    single text segment. Group messages prepend an ``@sender`` so the
    reply is clearly addressed (matches qqBot.js / Rust)."""
    if event.message_type == MessageType.GROUP:
        from corlinman_channels.onebot import AtSegment

        gid = event.group_id or 0
        return SendGroupMsg(
            group_id=gid,
            message=[
                AtSegment(qq=str(event.user_id)),
                TextSegment(text=f" {body}"),
            ],
        )
    return SendPrivateMsg(
        user_id=event.user_id,
        message=[TextSegment(text=body)],
    )


# ---------------------------------------------------------------------------
# Telegram channel
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class TelegramChannelParams:
    """Parameters for :func:`run_telegram_channel`. Mirrors Rust
    ``TelegramParams``."""

    config: Any
    """``cfg.channels.telegram`` — must expose ``bot_token``, optional
    ``allowed_chat_ids``, optional ``keyword_filter``, optional
    ``require_mention_in_groups``."""

    model: str = ""
    chat_service: ChatServiceLike | None = None
    base_url: str = "https://api.telegram.org"


async def run_telegram_channel(
    params: TelegramChannelParams,
    cancel: asyncio.Event,
) -> None:
    """Spawn the Telegram channel loop and run until ``cancel`` is set.

    Mirrors Rust ``run_telegram_channel`` in ``telegram/service.rs``.
    Inbound long-poll + outbound replies via :class:`TelegramSender`.
    """
    cfg = params.config
    bot_token = _attr(cfg, "bot_token", "")
    if not bot_token:
        raise ValueError("channels.telegram.bot_token is empty")

    tg_cfg = TelegramConfig(
        bot_token=str(bot_token),
        allowed_chat_ids=list(_attr(cfg, "allowed_chat_ids", []) or []),
        keyword_filter=list(_attr(cfg, "keyword_filter", []) or []),
        require_mention_in_groups=bool(_attr(cfg, "require_mention_in_groups", False)),
        base_url=str(_attr(cfg, "base_url", params.base_url)),
    )
    # The adapter owns its HTTP client (long-poll cadence is heavy);
    # the sender gets its own (short, eager-shutdown).
    adapter = TelegramAdapter(tg_cfg)
    send_client = httpx.AsyncClient()
    sender = TelegramSender(send_client, tg_cfg.bot_token, base=tg_cfg.base_url)
    pending: set[asyncio.Task[None]] = set()
    try:
        async with adapter:
            iterator = adapter.inbound()
            while not cancel.is_set():
                ev = await _race_iter_or_cancel(iterator, cancel)
                if ev is None:
                    break
                if params.chat_service is None:
                    continue
                t = asyncio.create_task(
                    handle_one_telegram(
                        params.chat_service,
                        ev,
                        params.model,
                        sender,
                        cancel,
                    )
                )
                pending.add(t)
                t.add_done_callback(pending.discard)
    finally:
        for t in pending:
            t.cancel()
        await send_client.aclose()


async def handle_one_telegram(
    chat_service: ChatServiceLike,
    inbound: InboundEvent[Any],
    model: str,
    sender: TelegramSender,
    cancel: asyncio.Event,
) -> None:
    """Run one Telegram chat turn and post the reply via
    :class:`TelegramSender`. Parallel structure to :func:`handle_one_qq`.
    """
    request = {
        "model": model,
        "messages": [{"role": "user", "content": inbound.text}],
        "session_key": inbound.binding.session_key(),
        "stream": True,
        "max_tokens": None,
        "temperature": None,
        "attachments": list(inbound.attachments),
        "binding": inbound.binding,
    }
    stream = await chat_service.run(request, cancel)
    text_parts: list[str] = []
    error_message: str | None = None
    async for ev in stream:
        kind = _event_kind(ev)
        if kind == "token_delta":
            text_parts.append(getattr(ev, "text", "") or "")
        elif kind == "done":
            break
        elif kind == "error":
            error_message = getattr(ev, "error", "") or getattr(ev, "message", "")
            break

    if error_message is not None:
        body = f"[corlinman error] {error_message}"
    else:
        body = "".join(text_parts)
        if not body.strip():
            return

    # ``inbound.binding.thread`` is the chat_id (Telegram thread = chat).
    chat_id = int(inbound.binding.thread)
    reply_to: int | None = None
    if inbound.message_id is not None:
        try:
            reply_to = int(inbound.message_id)
        except ValueError:
            reply_to = None
    await sender.send_message(chat_id, body, reply_to_message_id=reply_to)


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


async def _race_iter_or_cancel(
    iterator: AsyncIterator[Any],
    cancel: asyncio.Event,
) -> Any | None:
    """Get the next item from ``iterator`` or ``None`` if ``cancel``
    fires first. Equivalent of Rust ``tokio::select! { recv() => ...,
    cancelled() => break }``.
    """
    next_task = asyncio.create_task(iterator.__anext__())
    cancel_task = asyncio.create_task(cancel.wait())
    done, pending = await asyncio.wait(
        {next_task, cancel_task}, return_when=asyncio.FIRST_COMPLETED
    )
    for t in pending:
        t.cancel()
    if cancel_task in done:
        if next_task in done and not next_task.cancelled():
            # Race tie — both fired; consume the value we already got.
            try:
                return next_task.result()
            except (StopAsyncIteration, BaseException):
                return None
        return None
    try:
        return next_task.result()
    except StopAsyncIteration:
        return None


def _event_kind(ev: Any) -> str:
    """Best-effort discriminator extraction.

    Supports either ``ev.kind`` (string) or class-name fallbacks
    (``TokenDelta``, ``ToolCall``, ``Done``, ``Error``). Returns
    ``"unknown"`` for anything else."""
    k = getattr(ev, "kind", None)
    if isinstance(k, str):
        return k.lower()
    name = type(ev).__name__
    mapping = {
        "TokenDelta": "token_delta",
        "ToolCall": "tool_call",
        "Done": "done",
        "Error": "error",
        "InternalChatEvent": "token_delta",
    }
    return mapping.get(name, name.lower())


def _attr(obj: Any, name: str, default: Any) -> Any:
    """Walk attribute / mapping access uniformly. Tolerates both
    ``SimpleNamespace`` configs and TOML-loaded dicts."""
    if obj is None:
        return default
    if isinstance(obj, dict):
        return obj.get(name, default)
    return getattr(obj, name, default)


def _coerce_keywords(raw: Any) -> GroupKeywords:
    """Normalize a keyword map to ``dict[str, list[str]]``. Accepts
    either a dict from the loaded config or ``None``."""
    if not raw:
        return {}
    out: GroupKeywords = {}
    for k, v in raw.items():
        out[str(k)] = [str(x) for x in v]
    return out


# ---------------------------------------------------------------------------
# Re-export for the channel.py wrapper
# ---------------------------------------------------------------------------

#: ``corlinman_channels.channel.QqChannel`` imports this lazily; the
#: orchestration helpers above are the public surface.
_ = field  # keep dataclasses import alive for mypy
